File: engine/language-server/src/main/scala/org/enso/languageserver/boot/resource/ZioRuntimeInitialization.scala
package org.enso.languageserver.boot.resource

import akka.event.EventStream
import com.typesafe.scalalogging.LazyLogging
import org.enso.languageserver.effect
import org.enso.languageserver.event.InitializedEvent

import scala.concurrent.{ExecutionContext, Future}

/** Initialization of ZIO runtime.
  *
  * @param runtime the runtime to initialize
  * @param eventStream events stream
  * @param ec the execution context
  */
class ZioRuntimeInitialization(
  runtime: effect.Runtime,
  eventStream: EventStream
)(implicit
  ec: ExecutionContext
) extends InitializationComponent
    with LazyLogging {

  @volatile
  private var _isInitialized: Boolean = false

  override def isInitialized: Boolean = _isInitialized

  /** @inheritdoc */
  override def init(): Future[InitializationComponent.Initialized.type] =
    if (isInitialized) Future.successful(InitializationComponent.Initialized)
    else
      Future {
        logger.info("Initializing ZIO runtime...")
        runtime.init()
        logger.info("ZIO runtime initialized [{}].", runtime)
        _isInitialized = true
        eventStream.publish(InitializedEvent.ZioRuntimeInitialized)
        InitializationComponent.Initialized
      }
}

File: engine/language-server/src/main/java/org/enso/languageserver/boot/resource/ZioRuntimeInitialization.java
package org.enso.languageserver.boot.resource;

import akka.event.EventStream;
import java.util.concurrent.Executor;
import org.enso.languageserver.effect.Runtime;
import org.enso.languageserver.event.InitializedEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Initialization of ZIO runtime. */
public class ZioRuntimeInitialization extends LockedInitialization {

  private final Runtime runtime;
  private final EventStream eventStream;
  private final Logger logger = LoggerFactory.getLogger(this.getClass());

  /**
   * Create an instance of ZIO runtime initialization component.
   *
   * @param executor the executor that runs the initialization
   * @param runtime the runtime to initialize
   * @param eventStream the events stream
   */
  public ZioRuntimeInitialization(Executor executor, Runtime runtime, EventStream eventStream) {
    super(executor);
    this.runtime = runtime;
    this.eventStream = eventStream;
  }

  @Override
  public void initComponent() {
    logger.info("Initializing ZIO runtime...");
    runtime.init();
    logger.info("ZIO runtime initialized [{}].", runtime);
    eventStream.publish(InitializedEvent.ZioRuntimeInitialized$.MODULE$);
  }
}
